### PR TITLE
Add scoping structs for Win32 objects

### PR DESCRIPTION
**Pull request by JeremyKuhne**
This builds on what I did for ScreenDC some time ago. It creates wrapper ref structs for deterministic undoing of resource creation and selection. The goal is to help prevent misordering and leaked objects while also reducing nesting complexity.

Note that discards for the new style usings is a coming enhancement in C# that will make this even cleaner. If we move towards typing native handles this pattern will work well with that.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3442)

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
Ultimately I hope to get rid of the class based contexts we have (such as `DeviceContext`)- it should be possible to do so with a little iteration.

**Comment by weltkante** [comment]
> Ultimately I hope to get rid of the class based contexts we have (such as DeviceContext)- it should be possible to do so with a little iteration.

That would be awesome, I recently noticed while debugging leaked Controls that the class based contexts have issues with finalization order. For example `WindowsGraphicContext` sometimes owns a `DeviceContext` but when finalizing the order is nondeterministic and it can treat the `DeviceContext` as leaked instead of disposing it as part of `WindowsGraphicsContext`.

**Comment by JeremyKuhne** [comment]
> turning DeviceContext into a struct would probably solve some problems here.

I would expect so. I was experimenting with this last summer and made decent progress on this. This PR is the start of me trying to get my dangling wip cleaned up.

File: src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.CreateDcScope.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            public CreateDcScope(IntPtr hdc)
+            {
+                HDC = CreateCompatibleDC(hdc);
+            }
+
+            public static implicit operator IntPtr(CreateDcScope dcScope) => dcScope.HDC;
+
+            public void Dispose()
+            {
+                if (HDC != null)
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.CreateDcScope.cs:33]: @JeremyKuhne just noticed - `HDC` is a struct, why are we comparing it to `null`?

Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.CreateDcScope.cs:33]: Yeah, fixing. Not super harmful as we're not checking the result.

**Pull request merged**